nushell: 0.65.0 -> 0.67.0

shanesveller opened the pull request:
###### Description of changes

https://www.nushell.sh/blog/2022-07-26-nushell-0_66.html
https://github.com/nushell/nushell/releases/tag/0.66.1
https://github.com/nushell/nushell/releases/tag/0.66.2
https://github.com/nushell/nushell/compare/0.66.0...0.66.2

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [x] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [X] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

shanesveller committed:
nushell: 0.65.0 -> 0.66.2

https://www.nushell.sh/blog/2022-07-26-nushell-0_66.html
https://github.com/nushell/nushell/releases/tag/0.66.1
https://github.com/nushell/nushell/releases/tag/0.66.2
https://github.com/nushell/nushell/compare/0.66.0...0.66.2

winterqt commented:
So, a few things:
1. `cargo add` is no longer a part of `cargo-edit`, it's now directly integrated in Cargo.
2. `cargo-edit` v0.9.0's (the version that this modification was originally made with) `--offline` doesn't even seem to work. I added a random dependency that wasn't in `Cargo.lock` in `cargoUpdateHook` (with `--offline`) and it worked just fine.

I think we should just switch back to patching, or write a minimal tool to parse the TOML of `Cargo.lock`, pull the `zstd-sys` version, and then add it to `Cargo.toml`. Ideally the first one.

cc @Artturin, who originally added the `cargo-edit` usage.

fortuneteller2k commented on pkgs/shells/nushell/default.nix:
```diff
@@ -21,16 +21,16 @@
 
-  cargoSha256 = "sha256-YqtM/1p6oP0+E0rYSFPeCbof06E81eC2PZIwkU7J0I4=";
+  cargoSha256 = "sha256-4gYafGaQu+UWMQoQ9bf8Cm5rKZhO5ZbKDOitC4HXNdI=";
   # Since 0.34, nu has an indirect dependency on `zstd-sys` (via `polars` and
```
This whole section (`depsExtraArgs` and `cargoUpdateHook`) isn't really necessary anymore, builds fine without it.

`cargoSha256 = "sha256-4gYafGaQu+UWMQoQ9bf8Cm5rKZhO5ZbKDOitC4HXNdI=";`

And dataframe commands work fine:

![image](https://user-images.githubusercontent.com/20619776/181837102-3f01f37d-5c92-4951-a91a-3402682de84b.png)

winterqt replied:
`zstd-sys` still depends on the Git submodule.

fortuneteller2k replied:
The [original complaint](https://github.com/NixOS/nixpkgs/pull/130760#issuecomment-883317581) was that it failed to build without resorting to the `pkg-config` cargo patch, seems like that's not the case anymore.

winterqt replied:
You'd be right, the `zstd-sys` crate now properly includes the `zstd` source without needing to fetch the submodule. However, wouldn't it still be a good idea to use our `zstd` if possible, like it is here?

fortuneteller2k replied:
Hmm, maybe, but it isn't worth fighting the build system for it.

winterqt replied:
It's not really fighting the build system if we just have to flip a feature flag, IMO.

liketechnik replied:
[This patch](https://gist.github.com/liketechnik/bfa19583cdcd55e6778615dfb17488b7) (unfortunately comment attachments do not support diffs, so linked as a gist) added to `cargoPatches` allows to use the system zstd with nushell 0.66.2. 

It's a (minor) modification of the patch from the original issue linked above. I've changed to location of the zstd dependency, in order to workaround rust-lang/cargo#7880.

shanesveller replied:
I see other packages in nixpkgs using this pattern (e.g. `cargo-about` and `cargo-deny`), which seems like enough precedent that I am comfortable with it. I've incorporated @liketechnik 's patch in 1babd46a500ac7a6c2d59a0366a966a406500999. Please let me know if you'd like me to handle the attribution any differently, I'm fine with whatever.

I'd feel best if we hear from @Artturin about it as well?

winterqt replied:
> It's a (minor) modification of the patch from the original issue linked above. I've changed to location of the zstd dependency, in order to workaround [rust-lang/cargo#7880](https://github.com/rust-lang/cargo/issues/7880).

I can't reproduce this issue by adding the exact same line you did to the workspace's `Cargo.toml`, FWIW. (It doesn't really matter, but definitely interesting -- were you able to reproduce it, or were you just adding it there just in case?)

liketechnik replied:
Oops. You're observation is most likely correct. Apparently I got confused with multiple attempts to get this working, a) replacing the version number from the original version with a wildcard and b) moving it from the workspace to a specific crate. And I guess a) is the relevant fix and *not* moving it away from the workspace.

Not sure if it should be moved back into the workspace's `Cargo.toml`, then.

winterqt replied:
It doesn't really matter honestly, ideally it would be in the `Cargo.toml` so it's future proof to more `zstd-sys` dependents, but depending on how often upstream changes its dependencies, the patch will probably have to be regenerated at some point so it could be just done then. 

But now that I think about it, the dependency in the `Cargo.toml` should probably be the package which directly consumes the `pkg-config` feature flag, which is `zstd-sys`, as a matter of form/correctness. What do you think, am I just nitpicking here?

(Also, this patch will hopefully be made redundant in the future. When I get a chance, I'll send a patch upstream to allow setting this option, so we won't have to patch it at all for the next version. (Anyone else interested can PR this change upstream, I don't mind -- I'm just also happy to do it.))

shanesveller committed:
nushell: patch Cargo to use pkgconfig for zstd

Reverts the meaningful change in 5f3b928696e7c03f03d3cf2fe970e4bca47367ee.

https://github.com/NixOS/nixpkgs/pull/183712#discussion_r934027513

See also: cargo-about, cargo-deny

Co-authored-by: Florian Warzecha <liketechnik@disroot.org>

shanesveller commented:
Result of `nixpkgs-review pr 183712` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>nushell</li>
  </ul>
</details>

winterqt commented on pkgs/shells/nushell/default.nix:
```diff
@@ -21,32 +21,19 @@
+  cargoPatches = [ ./zstd-pkg-config.patch ];
 
   nativeBuildInputs = [ pkg-config ]
     ++ lib.optionals (withExtraFeatures && stdenv.isLinux) [ python3 ];
```
```suggestion
    ++ lib.optionals (withExtraFeatures && stdenv.isLinux) [ python3 ]
    ++ lib.optionals stdenv.isDarwin [ rustPlatform.bindgenHook ];
```

shanesveller replied:
Thanks for all of your help!

winterqt reviewed (changes requested):
`cargo-edit` can be removed from the `callPackage` arguments.

shanesveller committed:
nushell: Include Apple SDK for libproc.h / bindgen

https://github.com/nushell/nushell/commit/9b6b8172761cbeb9000dfef3c2db51fc22a56d8e
https://github.com/andrewdavidmackenzie/libproc-rs/releases/tag/v0.12.0
https://github.com/NixOS/nixpkgs/blob/fec8b61321a6597d0a9f93468970e6d2cdb896f9/pkgs/development/libraries/webkitgtk/default.nix#L161-L166
https://github.com/NixOS/nixpkgs/blob/fec8b61321a6597d0a9f93468970e6d2cdb896f9/pkgs/tools/system/btop/default.nix#L32-L37

shanesveller commented:
a778cafada011ad62229d1987eb085d732fd6110 (at the time of writing) was introduced to resolve an Intel-Mac-only build failure around `libproc.h`/`bindgen`/`libproc` crate. I'm open to suggestions that are less wacky!

shanesveller commented:
Result of `nixpkgs-review pr 183712` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>nushell</li>
  </ul>
</details>

winterqt commented:
1. I may be misunderstanding how we handle frameworks, at least on x86_64, but the Foundation input shouldn't be required (it is, though, and it seems like that may be a bug). I'll loop back on this when I get the chance to ask someone knowledgeable.
2. Would it perhaps be better/more consistent to use the same `runCommand` invocation and explanation comment that the other packages you linked do for libproc?

shanesveller committed:
nushell: 0.66.2 -> 0.67.0

https://www.nushell.sh/blog/2022-08-16-nushell-0_67.html

shanesveller commented:
> 2\. Would it perhaps be better/more consistent to use the same `runCommand` invocation and explanation comment that the other packages you linked do for libproc? (Is there a reason you didn't do that? Was it because yours was shorter or because it didn't work or something? (Just curious, yours does work :))

I agree, so I incorporated the implied suggestion in the version committed as 85681b372e96b071f6f85d0f63d9d38111ef2f8f.

shanesveller commented:
(I'll fix the fork URLs that are currently present in 85681b372e96b071f6f85d0f63d9d38111ef2f8f , that was accidental.)
